File: ra-idp/src/authentication/usecases/change-password.test.ts
import { describe, expect, it } from 'bun:test'
import { Argon2idPasswordHasher } from '../../../adapters/crypto/argon2id-password-hasher'
import { InMemoryPasswordHistoryRepository } from '../../../adapters/persistence/memory/password-history-repo'
import { InMemoryUserRepository } from '../../../adapters/persistence/memory/user-repo'
import type { BreachedPasswordChecker } from '../ports/breached-password-checker'
import type { DomainEvent, User } from '../../spec-bindings/schemas'
import {
  CurrentPasswordMismatchError,
  PasswordReuseError,
  UserNotFoundError,
  changePassword,
} from './change-password'
import { PASSWORD_POLICY, PasswordPolicyError } from './password-policy'

interface Harness {
  userRepo: InMemoryUserRepository
  historyRepo: InMemoryPasswordHistoryRepository
  passwordHasher: Argon2idPasswordHasher
  events: DomainEvent[]
  sub: string
}

async function setupUser(initialPassword: string): Promise<Harness> {
  const userRepo = new InMemoryUserRepository()
  const historyRepo = new InMemoryPasswordHistoryRepository()
  // Argon2id は slow なので低コストプロファイルでテストする (memoryCost KiB, timeCost)
  const passwordHasher = new Argon2idPasswordHasher(8, 1)
  const hash = await passwordHasher.hash(initialPassword)
  const user: User = {
    sub: 'user-1',
    preferred_username: 'alice',
    password_hash: hash,
    email_verified: false,
    mfa_enrolled: false,
    created_at: '2024-01-01T00:00:00.000Z',
    updated_at: '2024-01-01T00:00:00.000Z',
  }
  await userRepo.save(user)
  return { userRepo, historyRepo, passwordHasher, events: [], sub: user.sub }
}

describe('changePassword', () => {
  it('現パス一致 + ポリシー合格 + 履歴未一致なら hash を更新し PasswordChanged を emit', async () => {
    const h = await setupUser('demo-password-1234')
    const now = new Date('2025-06-10T00:00:00Z')

    const updated = await changePassword(
      { ...h, emit: (e) => h.events.push(e) },
      { sub: h.sub, current_password: 'demo-password-1234', new_password: 'fresh-pass-9182', now },
    )

    expect(await h.passwordHasher.verify('fresh-pass-9182', updated.password_hash)).toBe(true)
    const persisted = await h.userRepo.findBySub(h.sub)
    expect(persisted?.password_hash).toBe(updated.password_hash)
    expect(persisted?.updated_at).toBe(now.toISOString())
    expect(h.events).toEqual([
      { type: 'PasswordChanged', occurredAt: now.toISOString(), sub: h.sub },
    ])
    const recent = await h.historyRepo.recent(h.sub, PASSWORD_POLICY.historyDepth)
    expect(recent).toHaveLength(1)
    expect(await h.passwordHasher.verify('fresh-pass-9182', recent[0].encoded)).toBe(true)
  })

  it('現パスが一致しないと CurrentPasswordMismatchError', async () => {
    const h = await setupUser('demo-password-1234')
    await expect(
      changePassword(
        { ...h, emit: (e) => h.events.push(e) },
        { sub: h.sub, current_password: 'wrong', new_password: 'fresh-pass-9182' },
      ),
    ).rejects.toBeInstanceOf(CurrentPasswordMismatchError)
    expect(h.events).toEqual([])
  })

  it('新パスがポリシー違反なら PasswordPolicyError', async () => {
    const h = await setupUser('demo-password-1234')
    await expect(
      changePassword(
        { ...h, emit: (e) => h.events.push(e) },
        { sub: h.sub, current_password: 'demo-password-1234', new_password: 'short' },
      ),
    ).rejects.toBeInstanceOf(PasswordPolicyError)
    expect(h.events).toEqual([])
  })

  it('現在のパスワードと同じ新パスは再利用として PasswordReuseError', async () => {
    const h = await setupUser('demo-password-1234')
    await expect(
      changePassword(
        { ...h, emit: (e) => h.events.push(e) },
        {
          sub: h.sub,
          current_password: 'demo-password-1234',
          new_password: 'demo-password-1234',
        },
      ),
    ).rejects.toBeInstanceOf(PasswordReuseError)
    expect(h.events).toEqual([])
  })

  it('直近 historyDepth 件のいずれかと一致したら PasswordReuseError', async () => {
    const h = await setupUser('demo-password-1234')
    const passwords = [
      'pw-history-aaaa-1',
      'pw-history-bbbb-2',
      'pw-history-cccc-3',
      'pw-history-dddd-4',
      'pw-history-eeee-5',
    ]
    let current = 'demo-password-1234'
    for (const next of passwords) {
      await changePassword(
        { ...h, emit: (e) => h.events.push(e) },
        { sub: h.sub, current_password: current, new_password: next },
      )
      current = next
    }
    // historyDepth=5 件を回したのでこの時点では最初の demo-password-1234 は履歴外。
    // しかし直近 5 件の "pw-history-aaaa-1" は history に残るため、再選択は失敗する。
    await expect(
      changePassword(
        { ...h, emit: (e) => h.events.push(e) },
        { sub: h.sub, current_password: current, new_password: 'pw-history-aaaa-1' },
      ),
    ).rejects.toBeInstanceOf(PasswordReuseError)
  })

  it('history depth を超えた古いパスワードは再利用可能', async () => {
    const h = await setupUser('demo-password-1234')
    // depth=2 で usecase を実行し、3 世代回した後に最古を再利用可能なことを確認
    const opts = {
      ...h,
      emit: (e: DomainEvent) => h.events.push(e),
      historyDepth: 2,
    }
    await changePassword(opts, {
      sub: h.sub,
      current_password: 'demo-password-1234',
      new_password: 'rotate-aaaa-1234',
    })
    await changePassword(opts, {
      sub: h.sub,
      current_password: 'rotate-aaaa-1234',
      new_password: 'rotate-bbbb-1234',
    })
    await changePassword(opts, {
      sub: h.sub,
      current_password: 'rotate-bbbb-1234',
      new_password: 'rotate-cccc-1234',
    })
    // depth=2 で見るのは直近 2 件 (cccc, bbbb)。aaaa は履歴外で再利用可能。
    const reused = await changePassword(opts, {
      sub: h.sub,
      current_password: 'rotate-cccc-1234',
      new_password: 'rotate-aaaa-1234',
    })
    expect(await h.passwordHasher.verify('rotate-aaaa-1234', reused.password_hash)).toBe(true)
  })

  it('存在しない sub は UserNotFoundError', async () => {
    const h = await setupUser('demo-password-1234')
    await expect(
      changePassword(
        { ...h, emit: (e) => h.events.push(e) },
        { sub: 'ghost', current_password: 'demo-password-1234', new_password: 'fresh-pass-9182' },
      ),
    ).rejects.toBeInstanceOf(UserNotFoundError)
  })

  it('breachedPasswordChecker がヒットを返したら PasswordPolicyError([breached])', async () => {
    const h = await setupUser('demo-password-1234')
    const seen: string[] = []
    const breachedPasswordChecker: BreachedPasswordChecker = {
      async isBreached(plain) {
        seen.push(plain)
        return plain === 'leaked-password-9999'
      },
    }
    await expect(
      changePassword(
        { ...h, emit: (e) => h.events.push(e), breachedPasswordChecker },
        {
          sub: h.sub,
          current_password: 'demo-password-1234',
          new_password: 'leaked-password-9999',
        },
      ),
    ).rejects.toMatchObject({
      name: 'PasswordPolicyError',
      violations: ['breached'],
    })
    expect(seen).toEqual(['leaked-password-9999'])
    expect(h.events).toEqual([])
  })

  it('breachedPasswordChecker が false を返せば従来通り成功する', async () => {
    const h = await setupUser('demo-password-1234')
    const breachedPasswordChecker: BreachedPasswordChecker = {
      async isBreached() {
        return false
      },
    }
    const updated = await changePassword(
      { ...h, emit: (e) => h.events.push(e), breachedPasswordChecker },
      {
        sub: h.sub,
        current_password: 'demo-password-1234',
        new_password: 'fresh-pass-9182',
      },
    )
    expect(await h.passwordHasher.verify('fresh-pass-9182', updated.password_hash)).toBe(true)
  })

  it('bundled policy 違反があるときは breachedPasswordChecker を呼ばない (早期失敗)', async () => {
    const h = await setupUser('demo-password-1234')
    let called = false
    const breachedPasswordChecker: BreachedPasswordChecker = {
      async isBreached() {
        called = true
        return true
      },
    }
    await expect(
      changePassword(
        { ...h, emit: (e) => h.events.push(e), breachedPasswordChecker },
        { sub: h.sub, current_password: 'demo-password-1234', new_password: 'short' },
      ),
    ).rejects.toBeInstanceOf(PasswordPolicyError)
    expect(called).toBe(false)
  })
})

File: ra-idp-go/internal/authentication/usecases/change_password_test.go
package usecases

import (
	"context"
	"errors"
	"testing"
	"time"

	"ra-idp-go/internal/adapters/crypto"
	"ra-idp-go/internal/adapters/persistence/memory"
	"ra-idp-go/internal/spec"
)

func TestChangePasswordUpdatesHashAndEmitsEvent(t *testing.T) {
	t.Parallel()

	userRepo := memory.NewUserRepository()
	historyRepo := memory.NewPasswordHistoryRepository()
	hasher := crypto.NewArgon2idPasswordHasher()
	hash, err := hasher.Hash("demo-password-1234")
	if err != nil {
		t.Fatal(err)
	}
	now := time.Date(2025, 6, 10, 0, 0, 0, 0, time.UTC)
	user := &spec.User{
		Sub: "user-1", PreferredUsername: "alice", PasswordHash: hash,
		CreatedAt: now.Add(-time.Hour), UpdatedAt: now.Add(-time.Hour),
	}
	if err := userRepo.Save(context.Background(), user); err != nil {
		t.Fatal(err)
	}
	var events []spec.DomainEvent

	updated, err := ChangePassword(context.Background(), ChangePasswordDeps{
		UserRepo:            userRepo,
		PasswordHasher:      hasher,
		PasswordHistoryRepo: historyRepo,
		Emit: func(event spec.DomainEvent) {
			events = append(events, event)
		},
	}, ChangePasswordInput{
		Sub:             user.Sub,
		CurrentPassword: "demo-password-1234",
		NewPassword:     "fresh-pass-9182",
		Now:             now,
	})
	if err != nil {
		t.Fatalf("change password: %v", err)
	}
	if updated.UpdatedAt != now {
		t.Fatalf("updated_at=%s, want %s", updated.UpdatedAt, now)
	}
	ok, err := hasher.Verify("fresh-pass-9182", updated.PasswordHash)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("updated hash does not verify new password")
	}
	recent, err := historyRepo.Recent(context.Background(), user.Sub, PasswordPolicyHistoryDepth)
	if err != nil {
		t.Fatal(err)
	}
	if len(recent) != 1 {
		t.Fatalf("history entries=%d, want 1", len(recent))
	}
	ok, err = hasher.Verify("fresh-pass-9182", recent[0].Encoded)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("history hash does not verify new password")
	}
	if len(events) != 1 {
		t.Fatalf("events=%d, want 1", len(events))
	}
	if _, ok := events[0].(*spec.PasswordChanged); !ok {
		t.Fatalf("event type=%T, want *spec.PasswordChanged", events[0])
	}
}

func TestChangePasswordRejectsCurrentPasswordMismatch(t *testing.T) {
	t.Parallel()

	userRepo := memory.NewUserRepository()
	historyRepo := memory.NewPasswordHistoryRepository()
	hasher := crypto.NewArgon2idPasswordHasher()
	hash, err := hasher.Hash("demo-password-1234")
	if err != nil {
		t.Fatal(err)
	}
	if err := userRepo.Save(context.Background(), &spec.User{
		Sub: "user-1", PreferredUsername: "alice", PasswordHash: hash,
		CreatedAt: time.Now().UTC(), UpdatedAt: time.Now().UTC(),
	}); err != nil {
		t.Fatal(err)
	}

	_, err = ChangePassword(context.Background(), ChangePasswordDeps{
		UserRepo:            userRepo,
		PasswordHasher:      hasher,
		PasswordHistoryRepo: historyRepo,
	}, ChangePasswordInput{
		Sub:             "user-1",
		CurrentPassword: "wrong-password",
		NewPassword:     "fresh-pass-9182",
	})
	if !errors.Is(err, ErrCurrentPasswordMismatch) {
		t.Fatalf("err=%v, want current password mismatch", err)
	}
}

func TestChangePasswordHonorsTenantOverridePolicy(t *testing.T) {
	t.Parallel()

	userRepo := memory.NewUserRepository()
	historyRepo := memory.NewPasswordHistoryRepository()
	hasher := crypto.NewArgon2idPasswordHasher()
	hash, err := hasher.Hash("demo-password-1234")
	if err != nil {
		t.Fatal(err)
	}
	now := time.Date(2025, 6, 10, 0, 0, 0, 0, time.UTC)
	if err := userRepo.Save(context.Background(), &spec.User{
		Sub: "user-1", PreferredUsername: "alice", PasswordHash: hash,
		CreatedAt: now.Add(-time.Hour), UpdatedAt: now.Add(-time.Hour),
	}); err != nil {
		t.Fatal(err)
	}

	strict := PasswordPolicySnapshot{MinLength: 24, MaxLength: 128, HistoryDepth: 5}
	_, err = ChangePassword(context.Background(), ChangePasswordDeps{
		UserRepo:            userRepo,
		PasswordHasher:      hasher,
		PasswordHistoryRepo: historyRepo,
		Policy:              strict,
	}, ChangePasswordInput{
		Sub:             "user-1",
		CurrentPassword: "demo-password-1234",
		NewPassword:     "fresh-pass-9182",
		Now:             now,
	})
	var policyErr *PasswordPolicyError
	if !errors.As(err, &policyErr) {
		t.Fatalf("err=%v, want PasswordPolicyError under tenant strict policy", err)
	}
	if len(policyErr.Violations) == 0 || policyErr.Violations[0] != ViolationTooShort {
		t.Fatalf("violations=%v, want first too_short", policyErr.Violations)
	}
}

func TestChangePasswordRejectsPasswordReuse(t *testing.T) {
	t.Parallel()

	userRepo := memory.NewUserRepository()
	historyRepo := memory.NewPasswordHistoryRepository()
	hasher := crypto.NewArgon2idPasswordHasher()
	initialHash, err := hasher.Hash("demo-password-1234")
	if err != nil {
		t.Fatal(err)
	}
	now := time.Now().UTC()
	if err := userRepo.Save(context.Background(), &spec.User{
		Sub: "user-1", PreferredUsername: "alice", PasswordHash: initialHash,
		CreatedAt: now, UpdatedAt: now,
	}); err != nil {
		t.Fatal(err)
	}
	reusedHash, err := hasher.Hash("pw-history-aaaa-1")
	if err != nil {
		t.Fatal(err)
	}
	if err := historyRepo.Add(context.Background(), "user-1", reusedHash, now); err != nil {
		t.Fatal(err)
	}

	_, err = ChangePassword(context.Background(), ChangePasswordDeps{
		UserRepo:            userRepo,
		PasswordHasher:      hasher,
		PasswordHistoryRepo: historyRepo,
	}, ChangePasswordInput{
		Sub:             "user-1",
		CurrentPassword: "demo-password-1234",
		NewPassword:     "pw-history-aaaa-1",
	})
	if !errors.Is(err, ErrPasswordReused) {
		t.Fatalf("err=%v, want password reused", err)
	}
}
